### ISSUE TITLE
STDOUT/STDIN buffer size error when passing arguments into Windows PowerShell

### ISSUE THREAD

**Issue by KUTlime** [PowerShell/PowerShell#20775]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

```powershell
powershell {''} -Args ('x' * 100000)
```

For a minimal, reproducible example, see [this](https://github.com/PowerShell/PowerShell/issues/14775#issuecomment-779687941) comment. When I debug the PS run, it pointed me to STDOUT/STDIN buffer size too.

### Expected behavior

```console
No error
```

### Actual behavior

```console
An error upon call.
```

### Error details

```console
Exception             :
    Type           : System.Management.Automation.ApplicationFailedException
    ErrorRecord    :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Program 'powershell.exe' failed to run: The Process object must have the UseShellExecute property
 set to false in order to use environment variables.At line:1 char:1
                      + powershell {''} -Args ('x' * 100000)
                      + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
            HResult : -2146233087
        CategoryInfo          : ResourceUnavailable: (:) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : NativeCommandFailed
        InvocationInfo        :
            ScriptLineNumber : 1
            OffsetInLine     : 1
            HistoryId        : -1
            Line             : powershell {''} -Args ('x' * 100000)
            PositionMessage  : At line:1 char:1
                               + powershell {''} -Args ('x' * 100000)
                               + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            CommandOrigin    : Internal
        ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
    TargetSite     :
        Name          : InitNativeProcess
        DeclaringType : System.Management.Automation.NativeCommandProcessor, System.Management.Automation, Version=7.3.
9.500, Culture=neutral, PublicKeyToken=31bf3856ad364e35
        MemberType    : Method
        Module        : System.Management.Automation.dll
    Message        : Program 'powershell.exe' failed to run: The Process object must have the UseShellExecute property
set to false in order to use environment variables.At line:1 char:1
                     + powershell {''} -Args ('x' * 100000)
                     + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
    Data           : System.Collections.ListDictionaryInternal
    InnerException :
        Type       : System.InvalidOperationException
        TargetSite :
            Name          : StartWithShellExecuteEx
            DeclaringType : System.Diagnostics.Process
            MemberType    : Method
            Module        : System.Diagnostics.Process.dll
        Message    : The Process object must have the UseShellExecute property set to false in order to use environment
 variables.
        Source     : System.Diagnostics.Process
        HResult    : -2146233079
        StackTrace :
   at System.Diagnostics.Process.StartWithShellExecuteEx(ProcessStartInfo startInfo)
   at System.Management.Automation.NativeCommandProcessor.InitNativeProcess()
    Source         : System.Management.Automation
    HResult        : -2146233087
    StackTrace     :
   at System.Management.Automation.NativeCommandProcessor.InitNativeProcess()
   at System.Management.Automation.NativeCommandProcessor.Prepare(IDictionary psDefaultParameterValues)
   at System.Management.Automation.CommandProcessorBase.DoPrepare(IDictionary psDefaultParameterValues)
   at System.Management.Automation.Internal.PipelineProcessor.Start(Boolean incomingStream)
   at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input)
--- End of stack trace from previous location ---
   at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input)
   at System.Management.Automation.PipelineOps.InvokePipeline(Object input, Boolean ignoreInput, CommandParameterIntern
al[][] pipeElements, CommandBaseAst[] pipeElementAsts, CommandRedirection[][] commandRedirections, FunctionContext func
Context)
   at System.Management.Automation.Interpreter.ActionCallInstruction`6.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
CategoryInfo          : ResourceUnavailable: (:) [], ApplicationFailedException
FullyQualifiedErrorId : NativeCommandFailed
InvocationInfo        :
    ScriptLineNumber : 1
    OffsetInLine     : 1
    HistoryId        : -1
    Line             : powershell {''} -Args ('x' * 100000)
    PositionMessage  : At line:1 char:1
                       + powershell {''} -Args ('x' * 100000)
                       + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.9
PSEdition                      Core
GitCommitId                    7.3.9
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

![image](https://github.com/PowerShell/PowerShell/assets/8374320/b53763c0-de46-464d-8d4f-afc5ef97541f)

**Comment by mklement0** [comment]
Note that the problem is unrelated to stdin / stdout. It is _the length of the command line_ that is constructed behind the scenes, which exceeds what are ultimately _system_ limits.

The way that the problem surfaces with `powershell.exe` is curious, however. With `pwsh.exe` the error message provides a slightly better, but still confusing clue:

```none
… The filename or extension is too long. …
```

Only on  Unix-like platforms do you get a direct indication of what's wrong:

```none
… Argument list too long … 
```

(There's a formatting bug not shown here: The "At ..." position message is directly concatenated to the message above.)

The way PowerShell constructs the actual process command line when you call its CLI with a _script block_ actually makes the command line significantly _longer_ than the original, because it Base64-encodes both the script block and the - CLIXML-serialized - arguments, which you can verify as follows:

```powershell
pwsh { [Environment]::CommandLine } -Arg ('x' * 10)
```

---

In short: While the error messages can definitely be improved, the core problem is running into a system limitation.

There are two **workarounds** (note:  `& (Get-Process -Id $PID).Path` invokes the CLI of whichever PowerShell edition you're running):

* Use an _environment variable_:

```powershell
$env:FOO = 'x' * 100000
& (Get-Process -Id $PID).Path { $env:FOO.Length }
$env:FOO = $null
```

This works in PS _Core_ on Windows as well as on Unix-like platforms, but not in _Windows PowerShell_ (in this case the error message is helpful: "Environment variable name or value is too long.")

* Use _stdin_ (works in Windows PowerShell too, and is generally the better way to pass large values):

```powershell
'x' * 100000 | & (Get-Process -Id $PID).Path -nop { $($input).Length }
```
